### PR TITLE
Make sure the agent view is working with the Scheduler

### PR DESCRIPTION
**Pull request by hlloreda**
# Description

Updated a test to make sure the agent view is working with the Scheduler

closes https://github.com/inmanta/inmanta-core/issues/8186

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [ ] Attached issue to pull request
- [ ] Changelog entry
- [ ] Type annotations are present
- [ ] Code is clear and sufficiently documented
- [ ] No (preventable) type errors (check using make mypy or make mypy-diff)
- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

**Comment by hlloreda** [comment]
Data retrieved from the view:
```
[{'environment': '9f2158f8-b999-4b17-bbf7-048776497f81', 'name': '$__scheduler', 'last_failover': '2024-10-16T15:18:56.342750+00:00', 'paused': False, 'process_id': 'f5fc2fbc-8bd1-11ef-8330-42375fa4af91', 'process_name': 'fedora', 'unpause_on_resume': None, 'status': 'up'}, {'environment': '9f2158f8-b999-4b17-bbf7-048776497f81', 'name': 'agent1', 'last_failover': None, 'paused': False, 'process_id': None, 'process_name': None, 'unpause_on_resume': None, 'status': 'down'}]
```

File: tests/agent_server/deploy/e2e/test_autostarted.py [hunk 1]
```diff
@@ -688,6 +688,13 @@ def wait_for_new_executor() -> bool:
     for children in current_resumed_children.values():
         assert children.is_running()
 
+    result = await client.get_agents(environment)
+    assert result.code == 200
+    actual_data = result.result["data"]
+    assert len(actual_data) == 2
+    assert all([not agent["paused"] for agent in actual_data])
+    assert {e["name"] for e in actual_data} == {const.AGENT_SCHEDULER_ID, "agent1"}
```
Comment (wouterdb) [inline tests/agent_server/deploy/e2e/test_autostarted.py:696]: 1. I don't think we want to show the `AGENT_SCHEDULER_ID` 
2. I think we should try to set a meaningfull up/down  state (e.g. `up iff scheduler.state  == UP and not agent.paused`

**Review by wouterdb — CHANGES_REQUESTED** [review]

**Review by wouterdb — CHANGES_REQUESTED** [review]

File: tests/deploy/e2e/test_autostarted.py [hunk 2]
```diff
@@ -1131,3 +1200,152 @@ async def check_resource_in_state(expected_resources: int = 1, state: str = "dep
...
+        "environment": environment,
+        "last_failover": actual_data[0]["last_failover"],
+        "name": "agent1",
+        "paused": False,
+        "process_id": None,
+        "process_name": None,
+        "status": "down",
+        "unpause_on_resume": None,
+    }
+    assert actual_data[0] == expected_data
```
Comment (hlloreda) [inline tests/deploy/e2e/test_autostarted.py:1351]: @wouterdb  I added locally this part in the test (for testing purposes only):
```
    await asyncio.sleep(10)
    await retry_limited(
        wait_for_terminated_status,
        timeout=const.EXECUTOR_GRACE_HARD + 2,
        current_children=children_after_deployment.children,
        expected_terminated_process=2,
    )
    await retry_limited(
        wait_for_terminated_status,
        timeout=const.EXECUTOR_GRACE_HARD + 2,
        current_children=children_after_deployment.children,
        expected_terminated_process=3,
    )
```
But this fails because the fork server and the executor seem somehow to survive: 
```
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB post_mortem >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /home/hugo/Documents/Inmanta/copies/inmanta-core/tests/utils.py(87)retry_limited()
-> raise AssertionError("Bounded wait failed")
(Pdb) up
> /home/hugo/Documents/Inmanta/copies/inmanta-core/tests/deploy/e2e/test_autostarted.py(1354)test_scheduler_killed()
-> await retry_limited(
(Pdb) children_after_deployment.children
[psutil.Process(pid=29456, name='python', status='terminated', started='11:45:53'), psutil.Process(pid=29495, name='inmanta: multiprocessing fork server', status='sleeping', started='11:45:55'), psutil.Process(pid=29497, name='inmanta: executor process 78f726cec1879275a4cc2fa2c103afb5 - disconnected', status='sleeping', started='11:45:55')]
```
The executor seems dead but not cleaned, is this expected?

**Comment by hlloreda** [comment]
@wouterdb This test is failing: `test_agent_process_details_with_report` because it tries to fetch process details of a process that does not exist anymore (+ the fact that the scheduler is not returned from the agent view). How should I handle this? Is it still something we want to support (fetch process details) for scheduler processes?

File: src/inmanta/data/dataview.py [hunk 3]
```diff
@@ -1211,19 +1211,21 @@ def get_base_url(self) -> str:
...
+                                                ELSE 'up'
                                             END) as status""",
-            from_clause=f" FROM {Agent.table_name()} as a LEFT JOIN public.agentinstance ai ON a.id_primary=ai.id "
-            " LEFT JOIN public.agentprocess ap ON ai.process = ap.sid",
-            filter_statements=[" a.environment = $1 "],
-            values=[self.environment.id],
+            from_clause=f" FROM  (SELECT sched.id, sched.halted FROM {Environment.table_name()} AS sched "
+            "WHERE sched.id =$1) AS sched "
+            f"RIGHT JOIN {Agent.table_name()} a ON a.environment = sched.id "
+            f"RIGHT JOIN (SELECT schedagent.name, schedagent.id_primary FROM {Agent.table_name()} AS schedagent "
```
Comment (wouterdb) [inline src/inmanta/data/dataview.py:1222]:
```suggestion
            from_clause= f"RIGHT JOIN (SELECT schedagent.name, schedagent.id_primary FROM {Agent.table_name()} AS schedagent "
```

I think only the last bit is used?

Comment (wouterdb) [reply src/inmanta/data/dataview.py:1222]: Why a right join?

File: src/inmanta/data/dataview.py [hunk 4]
```diff
@@ -1195,7 +1196,6 @@ def __init__(
     def allowed_filters(self) -> dict[str, type[Filter]]:
         return {
             "name": ContainsPartialFilter,
-            "process_name": ContainsPartialFilter,
```
Comment (wouterdb) [inline src/inmanta/data/dataview.py:1198]: This will break the front-end. I would just make this filter do nothing.

File: src/inmanta/data/dataview.py [hunk 5]
```diff
@@ -1211,19 +1211,21 @@ def get_base_url(self) -> str:
...
                                             END) as status""",
-            from_clause=f" FROM {Agent.table_name()} as a LEFT JOIN public.agentinstance ai ON a.id_primary=ai.id "
-            " LEFT JOIN public.agentprocess ap ON ai.process = ap.sid",
-            filter_statements=[" a.environment = $1 "],
-            values=[self.environment.id],
+            from_clause=f" FROM  (SELECT sched.id, sched.halted FROM {Environment.table_name()} AS sched "
+            "WHERE sched.id =$1) AS sched "
+            f"RIGHT JOIN {Agent.table_name()} a ON a.environment = sched.id "
+            f"RIGHT JOIN (SELECT schedagent.name, schedagent.id_primary FROM {Agent.table_name()} AS schedagent "
+            f"WHERE schedagent.name ='{const.AGENT_SCHEDULER_ID}') AS schedagent ON a.name <> schedagent.name ",
```
Comment (wouterdb) [inline src/inmanta/data/dataview.py:1223]: 1. This 'AS' statement in here looks a bit weird to me
2. You have to make sure we select the scheduler from the right environment

File: tests/server/test_agents_api.py [hunk 6]
```diff
@@ -118,12 +118,12 @@ def agent_names(agents: list[dict[str, str]]) -> list[str]:
     return [agent["name"] for agent in agents]
 
 
-@pytest.mark.parametrize("order_by_column", ["name", "status", "process_name", "last_failover", "paused"])
```
Comment (wouterdb) [inline tests/server/test_agents_api.py:121]: if would be  interesting to make only the absolute minimal amount of modification to this testcase, so we know we don't break the front-end

**Comment by wouterdb** [comment]
> `test_agent_process_details_with_report`

Good question, I suppose you could report the process of the scheduler as the process of every agent and then it would all still work.

File: src/inmanta/data/dataview.py [hunk 7]
```diff
@@ -1211,16 +1213,19 @@ def get_base_url(self) -> str:
...
                                             END) as status""",
-            from_clause=f" FROM {Agent.table_name()} as a LEFT JOIN public.agentinstance ai ON a.id_primary=ai.id "
-            " LEFT JOIN public.agentprocess ap ON ai.process = ap.sid",
-            filter_statements=[" a.environment = $1 "],
-            values=[self.environment.id],
+            from_clause=f" FROM {Agent.table_name()} a "
+            "LEFT JOIN (SELECT sa.name, sa.id_primary, ap.hostname, ai.process FROM agent sa  "
+            f"LEFT JOIN {AgentInstance.table_name()} ai ON sa.id_primary=ai.id "
+            f"LEFT JOIN {AgentProcess.table_name()} ap ON ai.process = ap.sid  "
+            "WHERE sa.environment = $1 AND sa.name = $2) sa_join ON a.name <> sa_join.name ",
```
Comment (wouterdb) [inline src/inmanta/data/dataview.py:1226]: the join on `<>` I don't know about, my SQL is too weak. 
@Hugo-Inmanta can you check this out?

File: tests/server/test_agents_api.py [hunk 8]
```diff
@@ -123,7 +123,7 @@ def agent_names(agents: list[dict[str, str]]) -> list[str]:
 async def test_agents_paging(server, client, env_with_agents: None, environment: str, order_by_column: str, order: str) -> None:
     result = await client.get_agents(
         environment,
-        filter={"status": ["paused", "up"]},
```
Comment (wouterdb) [inline tests/server/test_agents_api.py:126]: Why is this changed?

**Review by wouterdb — APPROVED** [review]

File: src/inmanta/data/dataview.py (hunk 7 above)
Comment (Hugo-Inmanta) [reply src/inmanta/data/dataview.py:1226]: I think it is working as expected i.e. for every nominal agent we retrieve the info for the hostname and pid of the scheduler in the env.
@hlloreda  do I understand correctly that this `ON` clause is there just to have something to join on ? ie we would expect the same result if we passed `ON true` ? In this case a `CROSS JOIN` would be preferable I think

File: src/inmanta/data/dataview.py [hunk 9]
```diff
@@ -1211,16 +1213,19 @@ def get_base_url(self) -> str:
 
     def get_base_query(self) -> SimpleQueryBuilder:
         base = SimpleQueryBuilder(
-            select_clause="""SELECT a.name, a.environment, last_failover, paused, unpause_on_resume,
-                                            ap.hostname as process_name, ai.process as process_id,
-                                            (CASE WHEN paused THEN 'paused'
-                                                WHEN id_primary IS NOT NULL THEN 'up'
-                                                ELSE 'down'
+            select_clause="""SELECT a.name, a.environment, a.last_failover, a.paused, a.unpause_on_resume,
```
Comment (Hugo-Inmanta) [inline src/inmanta/data/dataview.py:1216]: Could you clean up the formatting a bit while you're at it, to improve readability ?

**Review by Hugo-Inmanta — CHANGES_REQUESTED** [review]
